compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/types/Types.java

```diff
@@ -0,0 +1,201 @@
+    public final TypeSymbol BYTE;
+    public final TypeSymbol COMPILATION_ERROR;
+    public final CompilerContext context;
+
+
+    private Types(CompilerContext context) {
+        context.put(TYPES_KEY, this);
+        this.context = context;
+        TypesFactory typesFactory = TypesFactory.getInstance(context);
+        this.symbolFactory = SymbolFactory.getInstance(context);
+        this.symbolTable = SymbolTable.getInstance(context);
+        this.packageCache = PackageCache.getInstance(context);
+
+        this.BOOLEAN = typesFactory.getTypeDescriptor(symbolTable.booleanType);
+        this.INT = typesFactory.getTypeDescriptor(symbolTable.intType);
+        this.FLOAT = typesFactory.getTypeDescriptor(symbolTable.floatType);
+        this.DECIMAL = typesFactory.getTypeDescriptor(symbolTable.decimalType);
+        this.STRING = typesFactory.getTypeDescriptor(symbolTable.stringType);
+        this.NIL = typesFactory.getTypeDescriptor(symbolTable.nilType);
+        this.XML = typesFactory.getTypeDescriptor(symbolTable.xmlType);
+        this.ERROR = typesFactory.getTypeDescriptor(symbolTable.errorType);
+        this.FUNCTION = typesFactory.getTypeDescriptor(symbolTable.invokableType);
+        this.FUTURE = typesFactory.getTypeDescriptor(symbolTable.futureType);
+        this.TYPEDESC = typesFactory.getTypeDescriptor(symbolTable.typeDesc);
+        this.HANDLE = typesFactory.getTypeDescriptor(symbolTable.handleType);
+        this.STREAM = typesFactory.getTypeDescriptor(symbolTable.streamType);
+        this.ANY = typesFactory.getTypeDescriptor(symbolTable.anyType);
+        this.ANYDATA = typesFactory.getTypeDescriptor(symbolTable.anydataType);
+        this.NEVER = typesFactory.getTypeDescriptor(symbolTable.neverType);
+        this.READONLY = typesFactory.getTypeDescriptor(symbolTable.readonlyType);
+        this.JSON = typesFactory.getTypeDescriptor(symbolTable.jsonType);
+        this.BYTE = typesFactory.getTypeDescriptor(symbolTable.byteType);
+        this.COMPILATION_ERROR = typesFactory.getTypeDescriptor(symbolTable.semanticError);
+    }
+
+    /**
+     * Lookup for the symbol of a user defined type within a given module. This would be considering type
+     * definitions, enums, enum members, and class definitions as valid user defined types when
+     * looking up. The module is determined by the provided org, module name and the version.
+     *
+     * @param org           The organization of the looking up module
+     * @param moduleName    The name of the looking up module
+     * @param version       The version of the looking up module
+     * @param typeDefName   The type definition
+     * @return The {@link Symbol} of the user defined type
+     */
+    public Optional<Symbol> getTypeByName(String org, String moduleName, String version, String typeDefName) {
+        PackageID packageID = new PackageID(Names.fromString(org), Names.fromString(moduleName),
+                Names.fromString(version));
```

Comment:
Just wondering whether we need to handle cases like user passing `null` 🤔

Reply:
@pubudu91, 
The `Names.fromString()` itself would be throwing an NPE if a `null` value is passed.
Should we add an exception here?

Reply:
That's because it expects non-null values as the input. So we should ensure it is non-null before passing it to `fromString()`. So I guess then the question is the behaviour of our API with regard to the constraints of the params. Do we also expect the params to be non-null? If so, best to throw an exception for such cases I guess.